### ISSUE TITLE
v2.19 and mysql dont work together

### ISSUE THREAD

**Issue by pitgub** [filesender/filesender#851]
Followed this guide - https://docs.filesender.org/v2.0/install/ - to install filesender with the branch master-filesender-2.19 . I get the following error when I run the php script in step 9. The error disappears when using  master-filesender-2.17.  -

Error at terminal after running php script in step 9 - 

Updating views mysql select t.*,0 as fileid,a.created as acreated,a.author_type,a.author_id,a.target_type,a.target_id,a.event,a.id as aid  from Transfers t, AuditLogs a  where  a.target_id=cast(t.id as varchar(255))  and target_type = 'Transfer'   UNION select t.*,0 as fileid,a.created as acreated,a.author_type,a.author_id,a.target_type,a.target_id,a.event,a.id as aid  from Transfers t, AuditLogs a, Files f  where  f.transfer_id=t.id     and a.target_id=cast(f.id as varchar(255))    and target_type = 'File'   
Error, Rolling database changes back....
 This should leave the database state as it was before you started the script
Encountered exception : database_access_failure, see logs for details (uid: 5f7d36645c175) ...

From the filesender logs - 

2020-10-07 04:26:55] [upgrade:info] fk SQL is ALTER TABLE Transfers ADD CONSTRAINT transfer_passwordencoding FOREIGN KEY (password_encoding)  REFERENCES DBConstantPasswordEncodings (id)   on delete cascade on update restrict ;
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException exception uid:5f7d36645c175] database_access_failure
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException trace uid:5f7d36645c175] #0 /opt/filesender/filesender/classes/utils/DatabaseMysql.class.php(86): DBI::__callStatic('exec', Array)
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException trace uid:5f7d36645c175] #1 /opt/filesender/filesender/classes/utils/Database.class.php(178): DatabaseMysql::createView('Transfers', 'transfersauditl...', 'select t.*,0 as...')
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException trace uid:5f7d36645c175] #2 /opt/filesender/filesender/scripts/upgrade/database.php(605): Database::createView('Transfers', 'transfersauditl...', 'select t.*,0 as...')
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException trace uid:5f7d36645c175] #3 {main}
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175] "SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'varchar(255))  and target_type = 'Transfer'   UNION select t.*,0 as fileid,a.cre' at line 1"
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175] Array
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175] (
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175]     [name] => exec
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175]     [args] => Array
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175]         (
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175]             [0] => CREATE OR REPLACE VIEW transfersauditlogsview as select t.*,0 as fileid,a.created as acreated,a.author_type,a.author_id,a.target_type,a.target_id,a.event,a.id as aid  from Transfers t, AuditLogs a  where  a.target_id=cast(t.id as varchar(255))  and target_type = 'Transfer'   UNION select t.*,0 as fileid,a.created as acreated,a.author_type,a.author_id,a.target_type,a.target_id,a.event,a.id as aid  from Transfers t, AuditLogs a, Files f  where  f.transfer_id=t.id     and a.target_id=cast(f.id as varchar(255))    and target_type = 'File'
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175]         )
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175]
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175] )
[2020-10-07 04:30:44] [upgrade:error] [DBIUsageException details uid:5f7d36645c175]

Hope this helps to fix the issue.

**Comment by monkeyiq** [comment]
There were some updates for early MariaDB versions (https://github.com/filesender/filesender/pull/833) included in release 2.20. If you continue to experience issues with release 2.20 please let me know and which MariaDB version so I can try to iron those out.

**Comment by WebSpider** [comment]
I'm currently seeing the same issue when upgrading from 2.15 to 2.18 on DB version 10.1.47-MariaDB-0ubuntu0.18.04.1, same SQL error from my logs:

```text
Updating views for transfersview
Updating views mysql select * , DATEDIFF(now(),created) as created_days_ago  , DATEDIFF(now(),expires) as expires_days_ago  , DATEDIFF(now(),made_available) as made_available_days_ago   , options LIKE '%encryption":true%' as is_encrypted  , (CASE WHEN password_version=1 THEN 'user' ELSE 'generated' END) as password_origin   from Transfers 
Updating views for transfersauthview
Updating views mysql select t.id as id,t.userid as userid,u.authid as authid,a.saml_user_identification_uid as user_id,t.made_available,t.expires,t.created  FROM Transfers t, UserPreferences u, Authentications a where t.userid = u.id and u.authid = a.id  
Updating views for transferssizeview
Updating views mysql select t.*,sum(f.size) as size from Transfers t, Files f  where  f.transfer_id=t.id   group by t.id  
Updating views for transfersrecipientview
Updating views mysql select t.*,r.email as recipientemail,r.id as recipientid from Transfers t, Recipients r  where  r.transfer_id=t.id  
Updating views for transfersfilesview
Updating views mysql select t.*,f.name as filename,f.size as filesize from Transfers t, Files f  where  f.transfer_id=t.id  
Updating views for transfersauditlogsview
Updating views mysql select t.*,0 as fileid,a.created as acreated,a.author_type,a.author_id,a.target_type,a.target_id,a.event,a.id as aid  from Transfers t, AuditLogs a  where  a.target_id=cast(t.id as varchar(255))  and target_type = 'Transfer'   UNION select t.*,0 as fileid,a.created as acreated,a.author_type,a.author_id,a.target_type,a.target_id,a.event,a.id as aid  from Transfers t, AuditLogs a, Files f  where  f.transfer_id=t.id     and a.target_id=cast(f.id as varchar(255))    and target_type = 'File'   
Error, Rolling database changes back....
 This should leave the database state as it was before you started the script
Encountered exception : database_access_failure, see logs for details (uid: 5f9ad7dee3c1d) ...
```
I will do some checking if the errors still occur if i roll my install to 2.20

**Comment by monkeyiq** [comment]
The cast changed in 2.20 to produce the following SQL statement. I now have MariaDB 10.2, 10.1 and 10.0 available in addition to a more recent one I use for testing. The below query (From FileSender 2.20 IIRC) executes on 10.0 so it should be OK there.

```
CREATE OR REPLACE VIEW transfersauditlogsview as select t.*,0 as fileid,a.created 
    as acreated,a.author_type,a.author_id,a.target_type,a.target_id,a.event,a.id as aid  from Transfers t, AuditLogs a  
    where  a.target_id=cast(t.id as char) and target_type = 'Transfer'  
 UNION select t.*,0 as fileid,a.created as acreated,a.author_type,a.author_id,a.target_type,a.target_id,a.event,a.id as aid  
   from Transfers t, AuditLogs a, Files f  
   where  f.transfer_id=t.id     and a.target_id=cast(f.id as char)   and target_type = 'File'
```

**Comment by monkeyiq** [comment]
This should be fixed in 2.20 and above.
